spec/unit/lib/wifi_user/use_cases/sponsor_users_spec.rb

```diff
@@ -165,6 +168,76 @@ def a_plural_sponsor_confirmation_request
     end
   end
 
+  context 'On Failing to send an SMS' do
+    let(:sponsor) { 'Cassandra <cassandra@gov.uk>' }
+    let(:success_sponsees) { [] }
+    let(:failed_sponsees) { %w(+447770000666) }
+    let(:sponsees) { success_sponsees + failed_sponsees }
+    let(:do_not_reply_id) { production_do_not_reply_id }
+    let(:formatted_failed_sponsees) { "* +447770000666" }
+
+    let(:send_sms_gateway) do
```

Comment:
You're right that this is a monster of a mock. 

A few thoughts:
- In general, I think it's the right decision to mock here - we have control over this collaborator and it's a unit spec.
- Is this mock doing too much? It's both _verifying what it receives_ and _giving specific responses_ - does that make reading it difficult?
- Is hiding this mock behind a level of abstraction valuable? It'll make the spec more readable at a high-level, but might make it harder to understand this specific part.

Reply:
would like to make clear that this double isn't being used as a mock. We aren't verifying any executions, and only apply special behaviour with certain fixed inputs (essentially it's a Fake, not a Mock).
It allows behaviour, however only to the same degree the mocks are above.

I'm going to lean on saying yes, the abstraction is very valuable.
It's contained to a context, and now each of the sub-contexts are responsible only for what they care about, which is to verify combinations of successful and unsuccessful calls are made.

I'm open to suggestions for how to do this better with the abstraction in place.
Perhaps the setup can be broken out into a function, and placed in the `before` block?

```rb
let(:send_sms_gateway) { double(execute: double(success: true)) }

before(:each) do
  set_send_sms_gateway_execution_branches(
    successful_numbers: success_sponsees, unsucessful_numbers: failed_sponsees
  )
end

def set_send_sms_gateway_execution_branches(successful_numbers:, unsucessful_numbers:)
  ...
end
```

Reply:
> Perhaps the setup can be broken out into a function, and placed in the before block? 
I like this